extends capabilities and tests for finalizers

entro-pi opened the pull request:
Tested: via `make test`

Let me know if I totally borked up by making copyObject part of the reconciler interface.
It seemed like the best way to get access to the namespace in the same way that
other functions were doing it.

entro-pi commented:
/assign @srampal 
And because we were talking about this, I'd like their input.
/assign @yiqigao217

rjbez17 commented:
Is this still a WIP?

entro-pi commented:
I'd like to get input before I mark it done. So...ready for review?

On Thu, Oct 10, 2019, 7:38 PM Ryan Bezdicek <notifications@github.com>
wrote:

> Is this still a WIP?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQZU4ILSFOK4BK4LCEDQN7RLBA5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEA6RTJA#issuecomment-540875172>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQ3CR52RST2Q2JDL3N3QN7RLBANCNFSM4I7UHGSA>
> .
>

entro-pi commented:
I'll mark that as ready for review in a second.

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +285,25 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+	defer r.Forest.Unlock()
+	ns := r.Forest.Get(nsNm)
+	if !ns.Exists() {
+		return nil
```
If you look at the callers for copyObject, you'll see that a lot of them use the returned value immediately after it comes out of this function, so returning nil will result in a segfault. So for this solution to work, you'd have to modify all the call sites. That's a good "code smell" that this functionality doesn't belong here :)

entro-pi replied:
My nose is training!  Hahaha.

I should have listened to my gut when I started by adding the receiver.

Live and learn!

entro-pi commented:
Thank you!

I'll hunt down where the decision process is made and make those changes
tonight.

On Fri, Oct 11, 2019, 7:00 AM Adrian Ludwin <notifications@github.com>
wrote:

> *@adrianludwin* commented on this pull request.
>
> /hold
> ------------------------------
>
> In incubator/hnc/api/v1alpha1/hierarchy_types.go
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142#discussion_r333999453>
> :
>
> > @@ -103,6 +104,9 @@ type Condition struct {
>  	// - "OBJECT_OVERRIDDEN": an object in this namespace has been overridden from its parent and will no longer be updated
>  	//
>  	// - "OBJECT_DESCENDANT_OVERRIDDEN": an object in this namespace is no longer being propagated because a propagated copy has been modified
> +	//
> +	// - "OBJECT_HAS_FINALIZER": The object has a finalizer type set, and should not be propagated.
>
> "has a finalizer set" - no need for the word "type"
> ------------------------------
>
> In incubator/hnc/pkg/controllers/object_controller.go
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142#discussion_r334002563>
> :
>
> > @@ -137,12 +138,12 @@ func (r *ObjectReconciler) ensureCorrectAncestry(ctx context.Context, log logr.L
>  	}
>
>  	// We found the object in an ancestor namespace. Is this object the same as the source?
> -	instCopied := copyObject(inst)
> +	instCopied := r.copyObject(src.GetNamespace(), inst)
>
> This isn't the right place for this logic. copyObject is called in
> multiple contexts: deciding *whether* to copy the object, and *actually*
> copying the object. We only want to set the conditions during the decision
> process; it doesn't make any sense to be setting it *while* actually
> propagating!
>
> In addition, the function we're in right now is called
> ensureCorrectAncestry. It would be surprising for this function to also
> do a check on the *inherent* properties of the object.
>
> So I think I'd like to add some method called something like
> ensureCanPropagate which gets called either before or after the call to
> ensureCorrectAncestry. If it fails, update should early-exit, and it
> should add the condition to the forest.
> ------------------------------
>
> In incubator/hnc/pkg/controllers/object_controller.go
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142#discussion_r334005986>
> :
>
> > @@ -284,14 +285,25 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
>  	return nil
>  }
>
> -func copyObject(inst *unstructured.Unstructured) *unstructured.Unstructured {
> +func (r *ObjectReconciler) copyObject(nsNm string, inst *unstructured.Unstructured) *unstructured.Unstructured {
> +	r.Forest.Lock()
> +	defer r.Forest.Unlock()
> +	ns := r.Forest.Get(nsNm)
> +	if !ns.Exists() {
> +		return nil
>
> If you look at the callers for copyObject, you'll see that a lot of them
> use the returned value immediately after it comes out of this function, so
> returning nil will result in a segfault. So for this solution to work,
> you'd have to modify all the call sites. That's a good "code smell" that
> this functionality doesn't belong here :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQYQ6R6BDBJ52MZZICDQOCBINA5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHWEGVY#pullrequestreview-300696407>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQ7OWMROAJRRDMWVI5LQOCBINANCNFSM4I7UHGSA>
> .
>

yiqigao217 commented on incubator/hnc/api/v1alpha1/hierarchy_types.go:
```diff
@@ -28,6 +28,7 @@ const (
 	CritAncestor              Code = "CRIT_ANCESTOR"
 	ObjectOverridden          Code = "OBJECT_OVERRIDDEN"
 	ObjectDescendantOverriden Code = "OBJECT_DESCENDANT_OVERRIDDEN"
+	ObjectHasFinalizer        Code = "OBJECT_HAS_FINALIZER"
```
We can probably rename the condition to ```ObjectInternal```, considering we are checking 3 fields for now ```DeletionGracePeriodSeconds```, ```DeletionTimestamp```,```Finalizers```. We may also check more object internal issues in the future in the ```ensureCanPropagate()``` function suggested by Adrian below.

entro-pi replied:
Done!

entro-pi commented:
Cool! I was unsure of the wording too.

I'm on it!

On Fri, Oct 11, 2019, 7:52 AM yiqigao217 <notifications@github.com> wrote:

> *@yiqigao217* requested changes on this pull request.
> ------------------------------
>
> In incubator/hnc/api/v1alpha1/hierarchy_types.go
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142#discussion_r334029711>
> :
>
> > @@ -28,6 +28,7 @@ const (
>  	CritAncestor              Code = "CRIT_ANCESTOR"
>  	ObjectOverridden          Code = "OBJECT_OVERRIDDEN"
>  	ObjectDescendantOverriden Code = "OBJECT_DESCENDANT_OVERRIDDEN"
> +	ObjectHasFinalizer        Code = "OBJECT_HAS_FINALIZER"
>
> We can probably rename the condition to ObjectInternal, considering we
> are checking 3 fields for now DeletionGracePeriodSeconds,
> DeletionTimestamp,Finalizers. We may also check more object internal
> issues in the future in the ensureCanPropagate() function suggested by
> Adrian below.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQ77KBIBUNLLFIAOIL3QOCHJBA5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHWOJEA#pullrequestreview-300737680>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQ7VT623VMAEBTQ5UYLQOCHJBANCNFSM4I7UHGSA>
> .
>

yiqigao217 commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +285,25 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
-	copied.SetDeletionTimestamp(nil)
-	copied.SetFinalizers(nil) // TODO: double-check this is the right thing to do?
+	if copied.GetDeletionGracePeriodSeconds() != nil {
+		ns.SetCondition(forest.Local, tenancy.ObjectHasFinalizer, "finalizer set")
```
For ```OBJECT_``` conditions, the key should be a string in the format of ```group/version/kind/namespace/name```. ```forest.Local``` is the key for locally-set namespace conditions, not object conditions.

You can find the documentation at #76

entro-pi replied:
I THINK I referenced this correctly.

entro-pi commented:
Excellent! Thank you!

On Fri, Oct 11, 2019, 8:53 AM yiqigao217 <notifications@github.com> wrote:

> *@yiqigao217* commented on this pull request.
> ------------------------------
>
> In incubator/hnc/pkg/controllers/object_controller.go
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142#discussion_r334060205>
> :
>
> >  	// Clear all irrelevant fields. cf https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#objectmeta-v1-meta
>  	copied.SetCreationTimestamp(metav1.Time{})
> -	copied.SetDeletionGracePeriodSeconds(nil)
> -	copied.SetDeletionTimestamp(nil)
> -	copied.SetFinalizers(nil) // TODO: double-check this is the right thing to do?
> +	if copied.GetDeletionGracePeriodSeconds() != nil {
> +		ns.SetCondition(forest.Local, tenancy.ObjectHasFinalizer, "finalizer set")
>
> For OBJECT_ conditions, the key should be a string in the format of
> group/version/kind/namespace/name. forest.Local is the key for
> locally-set namespace conditions, not object conditions.
>
> You can find the documentation at #76
> <https://github.com/kubernetes-sigs/multi-tenancy/issues/76>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQ4QWJBEAM7VQMZ7CT3QOCOPXA5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHWYBLQ#pullrequestreview-300777646>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQYMTNVEK27WVHECUUDQOCOPXANCNFSM4I7UHGSA>
> .
>

entro-pi commented on incubator/hnc/api/v1alpha1/hierarchy_types.go:
```diff
@@ -103,6 +104,9 @@ type Condition struct {
 	//
 	// - "OBJECT_DESCENDANT_OVERRIDDEN": an object in this namespace is no longer being propagated because a propagated copy has been modified
+	//
+	// - "OBJECT_HAS_FINALIZER": The object has a finalizer type set, and should not be propagated.
```
Done

entro-pi commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -137,12 +138,12 @@ func (r *ObjectReconciler) ensureCorrectAncestry(ctx context.Context, log logr.L
 
 	// We found the object in an ancestor namespace. Is this object the same as the source?
-	instCopied := copyObject(inst)
+	instCopied := r.copyObject(src.GetNamespace(), inst)
```
Added the method mentioned and called it before the call to ensureCorrectAncestry, as well, undid the changes to the copyObject references.

entro-pi commented:
Ok! Made all changes requested, passes make test.

Happy Friday!

adrianludwin commented on incubator/hnc/api/v1alpha1/hierarchy_types.go:
```diff
@@ -103,6 +104,9 @@ type Condition struct {
 	//
 	// - "OBJECT_DESCENDANT_OVERRIDDEN": an object in this namespace is no longer being propagated because a propagated copy has been modified
+	//
+	// - "OBJECT_INTERNAL": The object has a finalizer set, and should not be propagated.
```
@yiqigao217 , what's the rationale behind this name? I'm not sure what "internal" means. Maybe something like OBJECT_METADATA with a description "This object has metadata that prevents it from being propagated to descendant namespaces" and the `msg` field will identify the actual problems?

yiqigao217 replied:
SGTM. Yes, metadata is better.

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -104,6 +105,12 @@ func (r *ObjectReconciler) SyncNamespace(ctx context.Context, log logr.Logger, n
 // to any child namespaces.
 func (r *ObjectReconciler) update(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) error {
+	//Checks to see if the object is correct and should propagate
+	if canPropagate := r.ensureCanPropagate(inst.GetNamespace(), inst); !canPropagate {
```
No need to introduce a new variable here:

`if !r.ensureCanPropagate(inst.GetNamespace(), inst) {`

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -104,6 +105,12 @@ func (r *ObjectReconciler) SyncNamespace(ctx context.Context, log logr.Logger, n
+	//Checks to see if the object is correct and should propagate
+	if canPropagate := r.ensureCanPropagate(inst.GetNamespace(), inst); !canPropagate {
+		log.Info("Object cannot propagate")
+		return fmt.Errorf("Object has finalizers set")
```
Returning an error means "the reconciler failed and needs to be retried." That's not what's happened here (the reconciler successfully determined that the object couldn't be propagated) so this should be `return nil`.

entro-pi replied:
I KNEW something was off when I had to learn a new way of sending an error!

I think I'm starting to understand code smell, now I just need to spot it more.

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +291,31 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+		return false
+	}
+	if inst.GetDeletionGracePeriodSeconds() != nil {
+		ns.SetCondition(fmt.Sprint(r.GVK, ns, nsNm), tenancy.ObjectInternal, "finalizer set")
```
A few things here:

- The key needs to be separated by slashes (`group/version/kind/namespace/name`), does your code do this? My reading of `fmt.Sprint` is that is separates the args using spaces.
- Not `"finalizer set"` - more like `"object is being deleted (deletionGracePeriodSeconds is set)"`

entro-pi replied:
It does not. Again, should have asked when I thought of it.

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +291,31 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+	if !ns.Exists() {
+		return false
+	}
+	if inst.GetDeletionGracePeriodSeconds() != nil {
```
I don't think this will compile - this doesn't return a pointer so it can't be compared to `nil`. See https://github.com/kubernetes-sigs/multi-tenancy/blob/master/incubator/hnc/pkg/controllers/hierarchy_controller.go#L116 for the correct way to check this.

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +291,31 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+		return false
+	}
+	if inst.GetDeletionTimestamp() != nil {
+		ns.SetCondition(fmt.Sprint(r.GVK, ns, nsNm), tenancy.ObjectInternal, "finalizer set")
```
Ditto. Also, rather than cutting-and-pasting the complex `fmt.Sprint` line, can you factor that out somehow?

entro-pi replied:
Factor that out meaning declare it as a local variable and use that as opposed to calling it all on one line and making it hard to read? Sure!

adrianludwin replied:
Yup, that works!

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +291,31 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+		return false
+	}
+	if inst.GetFinalizers() != nil {
+		ns.SetCondition(fmt.Sprint(r.GVK, ns, nsNm), tenancy.ObjectInternal, "finalizer set")
```
In this case `"finalizer set"` is technically correct, but the message is going to a human so a longer message is probably friendlier (eg `"object has finalizers set and can't be propagated"`)

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +291,31 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
 func copyObject(inst *unstructured.Unstructured) *unstructured.Unstructured {
 	copied := inst.DeepCopy()
-
-	// Clear all irrelevant fields. cf https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#objectmeta-v1-meta
```
Why delete this comment?

entro-pi replied:
Accidental, was writing lines around it and missed it in review.

adrianludwin replied:
thx

adrianludwin commented:
/hold

entro-pi commented:
Fyewf! That'll keep me busy for a while. Thank you for your input and
guidance!

(The comment deletion was accidental and I missed it in review.)

On Sun, Oct 13, 2019, 5:24 AM Adrian Ludwin <notifications@github.com>
wrote:

> /hold
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQZZIQUANAR3PAFF4J3QOMHOTA5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBCVEYQ#issuecomment-541413986>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQ2CUEWZ72FX4A7552LQOMHOTANCNFSM4I7UHGSA>
> .
>

entro-pi commented:
Having a bit of trouble with the unit test, reading documentation. Will report more as I finish this today.

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller_test.go:
```diff
@@ -4,13 +4,12 @@ import (
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/types"
-
-	tenancy "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
```
Can you restore this please? Usually we have (up to) three blocks of imports:

- Core libraries (eg `context`)
- Packages from other repos (eg `k8s.io/...`)
- Other packages from this repo (eg `v1alpha1`)

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller_test.go:
```diff
@@ -100,6 +99,15 @@ var _ = Describe("Secret", func() {
+		setParent(ctx, quxName, barName)
+		if hasAnyCondition(ctx, barName) {
+			panic("CONDITION SET")
+		}
```
Based on https://onsi.github.io/gomega/#making-assertions, I'd say something like:

`Expect(hasAnyCondition(ctx, barName)).ToNot(Equal(true))`

You shouldn't panic since it stops the rest of the tests from running.

Usually, you'll use `Eventually()` instead of `Expect()` since the latter runs immediately while the former can retry for up to two seconds (configurable in `suite_test.go` IIRC, but please don't change it). This is necessary because the controller is running asynchronously alongside the test code, so when you make a change (like updating the test apiserver), it takes a few moments to respond.

entro-pi commented:
Nice! Thank you. I figured Eventually had something to do with asynchronous
calls.

I'll restore the import and rewrite this test.

Do you have any ideas on testing for propagation as well? Maybe set a
condition, then a parent and then grab the object and test for a condition?
The step I'm having the most trouble with is grabbing the correct namespace
so I can test for a condition.


On Sun, Oct 13, 2019, 2:27 PM Adrian Ludwin <notifications@github.com>
wrote:

> *@adrianludwin* commented on this pull request.
> ------------------------------
>
> In incubator/hnc/pkg/controllers/object_controller_test.go
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142#discussion_r334297672>
> :
>
> > -
> -	tenancy "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
>
> Can you restore this please? Usually we have (up to) three blocks of
> imports:
>
>    - Core libraries (eg context)
>    - Packages from other repos (eg k8s.io/...)
>    - Other packages from this repo (eg v1alpha1)
>
> ------------------------------
>
> In incubator/hnc/pkg/controllers/object_controller_test.go
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142#discussion_r334297766>
> :
>
> > +		if hasAnyCondition(ctx, barName) {
> +			panic("CONDITION SET")
> +		}
>
> Based on https://onsi.github.io/gomega/#making-assertions, I'd say
> something like:
>
> Expect(hasAnyCondition(ctx, barName)).ToNot(Equal(true))
>
> You shouldn't panic since it stops the rest of the tests from running.
>
> Usually, you'll use Eventually() instead of Expect() since the latter
> runs immediately while the former can retry for up to two seconds
> (configurable in suite_test.go IIRC, but please don't change it). This is
> necessary because the controller is running asynchronously alongside the
> test code, so when you make a change (like updating the test apiserver), it
> takes a few moments to respond.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQ3EXUFSFEVVI7CG3BDQOOHELA5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCHY65SI#pullrequestreview-301067977>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQ746PVQ3AQGSFRURB3QOOHELANCNFSM4I7UHGSA>
> .
>

adrianludwin commented:
It should be relatively easy to set a condition in a test - just set the parent of a namespace to something that doesn't exist. There should be lots of examples in hierarchy_controller_test.go.

I'm not sure what you mean by "grabbing the correct namespace." Perhaps have a live chat or video meeting with @yiqigao217 and then pull me in if you're still having trouble?

entro-pi commented:
I think I've got it figured out. I'll get back at it tomorrow after work.

Cheers,
Thanks for all your help!

On Mon, Oct 14, 2019, 6:02 PM Adrian Ludwin <notifications@github.com>
wrote:

> It should be relatively easy to set a condition in a test - just set the
> parent of a namespace to something that doesn't exist. There should be lots
> of examples in hierarchy_controller_test.go.
>
> I'm not sure what you mean by "grabbing the correct namespace." Perhaps
> have a live chat or video meeting with @yiqigao217
> <https://github.com/yiqigao217> and then pull me in if you're still
> having trouble?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQ67XM23EZGFBNOHHPLQOUJCBA5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBHB6WY#issuecomment-541990747>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQ46KA7ZWQ5W5OIVHPTQOUJCBANCNFSM4I7UHGSA>
> .
>

yiqigao217 commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +292,38 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+	var conditionString string
+	if len(strings.Split(r.GVK.String(), ".")) > 1 {
+		affected := strings.Split(r.GVK.String(), ".")
+		conditionString = fmt.Sprint(affected[0], "/", affected[1], "/", affected[2], "/", nsNm, "/", ns.Name())
```
The last piece in the string should be ```inst.GetName()``` since it's the name of the object not namespace name. ```ns.Name()``` is the same as ```nsNm``` here.

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +292,38 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+		return false
+	}
+	//TODO TAYLOR
+	var conditionString string
```
`String` isn't a very descriptive suffix for a variable of type `string`. Perhaps `conditionKey` is more along the lines of what you meant?

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +292,38 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+	}
+	//TODO TAYLOR
+	var conditionString string
+	if len(strings.Split(r.GVK.String(), ".")) > 1 {
```
I don't understand this line? There will _always_ be a GVK for _every_ object. However, for legacy reason, the "core" objects (like `Pod` or `Secret`) have an empty string as the group. Eg, a pod "bar" in namespace "foo" should have the key `/v1/pod/foo/bar`, while a replica set "rs" in the same namespace would have the key `apps/v1/pod/foo/rs`.

adrianludwin commented on incubator/hnc/pkg/controllers/object_controller.go:
```diff
@@ -284,14 +292,38 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
+	//TODO TAYLOR
+	var conditionString string
+	if len(strings.Split(r.GVK.String(), ".")) > 1 {
+		affected := strings.Split(r.GVK.String(), ".")
```
Instead of splitting and combining, why not just use the components directly?

yiqigao217 commented:
> Do you have any ideas on testing for propagation as well? Maybe set a condition, then a parent and then grab the object and test for a condition? The step I'm having the most trouble with is grabbing the correct namespace so I can test for a condition.

I don't think you need to set a condition to test the propagation. The ```OBJECT_METADATA``` condition should be set by your code in object controller if anything's wrong with the object metadata. You can just test in 2 steps:
* Set finalizer (or deletion meta) to a src object and see if ```OBJECT_METADATA``` condition is set.
* Set a child namespace to the src object namespace and see if the src object is propagated to this child namespace.

entro-pi commented:
Thank you!

That actually answers all of my hanging questions.

I'll update tonight with more details and my rational.


On Tue, Oct 15, 2019, 8:59 AM yiqigao217 <notifications@github.com> wrote:

> Do you have any ideas on testing for propagation as well? Maybe set a
> condition, then a parent and then grab the object and test for a condition?
> The step I'm having the most trouble with is grabbing the correct namespace
> so I can test for a condition.
>
> I don't think you need to set a condition to test the propagation. The
> OBJECT_METADATA condition should be set by your code in object controller
> if anything's wrong with the object metadata. You can just test in 2 steps:
>
>    - Set finalizer (or deletion meta) to a src object and see if
>    OBJECT_METADATA condition is set.
>    - Set a child namespace to the src object namespace and see if the src
>    object is propagated to this child namespace.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQZMQ3NCVJCSUITIGGTQOXSEZA5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBJJXGQ#issuecomment-542284698>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQ6KM4HP37QLUC2Z4J3QOXSEZANCNFSM4I7UHGSA>
> .
>

entro-pi commented:
Ooh, I was confused about that for sure. I'll change the calls as it isn't
necessary to pass the namespace if it's passed with the instance.

On Tue, Oct 15, 2019, 8:21 AM yiqigao217 <notifications@github.com> wrote:

> *@yiqigao217* commented on this pull request.
> ------------------------------
>
> In incubator/hnc/pkg/controllers/object_controller.go
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142#discussion_r335019206>
> :
>
> > @@ -284,14 +292,38 @@ func (r *ObjectReconciler) onDelete(ctx context.Context, log logr.Logger, nnm ty
>  	return nil
>  }
>
> +func (r *ObjectReconciler) ensureCanPropagate(nsNm string, inst *unstructured.Unstructured) bool {
> +	r.Forest.Lock()
> +	defer r.Forest.Unlock()
> +	ns := r.Forest.Get(nsNm)
> +	if !ns.Exists() {
> +		return false
> +	}
> +	//TODO TAYLOR
> +	var conditionString string
> +	if len(strings.Split(r.GVK.String(), ".")) > 1 {
> +		affected := strings.Split(r.GVK.String(), ".")
> +		conditionString = fmt.Sprint(affected[0], "/", affected[1], "/", affected[2], "/", nsNm, "/", ns.Name())
>
> The last piece in the string should be inst.GetName() since it's the name
> of the object not namespace name. ns.Name() is the same as nsNm here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/multi-tenancy/pull/142?email_source=notifications&email_token=AM6RRQ5BPOCCDI4Q5W7BPS3QOXNV3A5CNFSM4I7UHGSKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCIABJLY#pullrequestreview-301995183>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM6RRQ7IPSEVM22Y6FYRDKLQOXNV3ANCNFSM4I7UHGSA>
> .
>

entro-pi commented:
I will continue work on this tomorrow night. Had some stuff come up.

adrianludwin commented:
Closing as @entro-pi told me they aren't ready to work on this for the time being.
